```{r}
library(SingleCellExperiment)
library(SuperCell)
library(SuperCellBM)
```

```{r config files}
source("./examples/config/Tian_config.R")
```

```{r flags}
ToComputeSC <- T
ToComputeSC_GE <- T

ToTestPackage <- F # @Loc, This is just to test whether package works (on reduced set of graining levels and seeds, turn it to FALSE to get real data @ all grainig levels and seeds)

filename_suf <- "" # variable to add a suffix to the saved files in case of testing of the package
```
# Set gamma
```{r}
gamma.seq.mc <- c(1, 2, 3, 4, 5, 6, 10)
gamma.seq <- c(1, 2, 5, 10, 20, 50, 100)
seed.seq <- .seed.seq[1:3]
```


## Get and set the main variables
Such as single-cell gene expression (`sc.GE`), single-cell counts (`sc.counts`), 
number of single cells (`N.c`) and total number of genes (`N.g`).
Set matrix column names to cellIDs (`cell.ids`) and row names to gene names (`gene.names`).
```{r get main values from the dataset}
preprocessed <- readRDS("~/switchdrive/Institution/CCB/superimpute/output/2021-08-18/object/preprocessed.Rds")
sc.counts <- preprocessed$sc_counts
sc.GE <- preprocessed$sc_GE

## this is not needed at this point, but will be used later
GT.cell.type <- preprocessed$sc_celllines
names(GT.cell.type) <- colnames(sc.counts)
N.clusters         <- length(unique(GT.cell.type))

GT.cell.type.names          <- names(table(GT.cell.type))
GT.cell.type.2.num          <- 1:length(unique(GT.cell.type))
names(GT.cell.type.2.num)   <- GT.cell.type.names
GT.cell.type.num            <- GT.cell.type.2.num[GT.cell.type]
names(GT.cell.type.num)     <- names(GT.cell.type)
```



## Compute Super-cells structure
for the Exact, Aprox (Super-cells obtained with the exact or approximate coarse-graining), Subsampling or Random (random grouping of cells into super-cells).

```{r compute super-cell structure, message=FALSE, warning=FALSE}
filename <- paste0('initial', filename_suf)

SC.list <- compute_supercells(
  sc.GE,
  ToComputeSC = ToComputeSC,
  data.folder = data.folder,
  filename = filename,
  gamma.seq = gamma.seq,
  n.var.genes = .N.var.genes,
  k.knn = .k.knn,
  n.pc = .N.comp,
  approx.N = .approx.N,
  fast.pca = TRUE,
  genes.use = .genes.use, 
  genes.exclude = .genes.omit,
  seed.seq = .seed.seq[1]
  )

cat(paste("Super-cell computed for:", paste(names(SC.list), collapse = ", "), 
          "\nat graining levels:", paste(names(SC.list[['Approx']]), collapse = ", "),
          "\nfor seeds:", paste(names(SC.list[['Approx']][[1]]), collapse = ", "), "\n",
          "\nand saved to / loaded from", paste0(filename, ".Rds")))
```

## Compute metacells in two settings:
- 'metacell_default' - when metacell is computed with the default parameters (from the tutorial), using gene set filtered by MC
- 'metacell_SC_like' - when metacell is computed with the same default parameters, but at the same set of genes as Super-cells 

```{r compute metacells, results='hide', message=FALSE}
SC.mc <- compute_supercells_metacells(
  sc.counts = sc.counts, 
  gamma.seq = gamma.seq.mc,
  SC.list = SC.list,
  proj.name = proj.name,
  ToComputeSC = TRUE, 
  mc.k.knn = 100,
  T_vm_def = 0.08,
  MC.folder = "MC", 
  MC_gene_settings = c('metacell_default', 'metacell_SC_like') # do not change
)
```

### Get actual graining levels obtained with Metacell
```{r get actual graining levels obtained with metacell}
additional_gamma_seq <- get_actual_gammas_metacell(SC.mc)

cat(paste("Metacells were computed in", length(names(SC.mc)), "settings:", paste(names(SC.mc), collapse = ", "), 
          "\nfor Gammas:", paste(names(SC.mc[[1]]), collapse = ", "), 
          "\nbut actual gammas are:", paste(additional_gamma_seq, collapse = ", ")
))


# manually expand MC because later we will have 2 different setting for MC profile: fp - footpring of MC, av - averaged 
SC.mc.fp <- SC.mc
names(SC.mc.fp) <- sapply(names(SC.mc), FUN = function(x){paste0(x, '_fp')})

SC.mc.av <- SC.mc
names(SC.mc.av) <- sapply(names(SC.mc), FUN = function(x){paste0(x, '_av')})

SC.mc.expanded <- c(SC.mc.fp, SC.mc.av)

names(SC.mc.expanded)
rm(SC.mc.fp, SC.mc.av, SC.mc)
```

### Concatenate Metacells to the list of Super-cells

```{r concat super-cell and metacells}
# SC.list <- c(SC.list, SC.mc.expanded)
# rm(SC.mc.expanded)
# 
# filename <- paste0("all", filename_suf)
# saveRDS(SC.list, file = file.path(data.folder, "SC", paste0(filename, ".Rds")))
# 
# cat(paste(
#   "Metacell data added to SC.list \nand now it contains:",
#   paste(names(SC.list), collapse = ", "),
#   "\nSC.list was saved to", file.path(data.folder, "SC", paste0(filename, ".Rds"))
# ))
```
Save metacell results
```{r}
saveRDS(SC.mc.expanded, "./output/MC_for_imputation.rds")
saveRDS(SC.list, "./output/SC_for_imputation.rds")
```


